### PR TITLE
Feature: Implement favourites menu

### PR DESCRIPTION
**Pull request by wutschel**
## Description

Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/268.
Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/55.

This PR implements support for Kodi´s favourites. Details:
- Adding a main menu entry named "Favourites"
- Loading favourites from Kodi and display name and -- if available and loadable -- thumbs
- When selecting an item either play (`media` type), activate window on Kodi server (`window` type) or show an error (`unknown` type)

Remarks:
- Not possible to start showing favourite pictures as they are reported as `unknown` type 
- TV station logo background selection does not work as the API just reports channels as `media` type (like movies or music)
- Change is made on top of https://github.com/xbmc/Official-Kodi-Remote-iOS/pull/283 for ease of changes to `AppDelegate.m`
- Needs to be aligned with the PRs related icon update coding style & refactoring and configurable main menu.

Screenshots:
https://abload.de/img/simulatorscreenshot-i7ekrd.png
https://abload.de/img/simulatorscreenshot-iz5k48.png
https://abload.de/img/simulatorscreenshot-iidj42.png

## Summary for release notes

Feature: Support Kodi favourites

### CODE REVIEW

**Comment by wutschel** [comment]
Rebased to master.

**Comment by wutschel** [comment]
Rebased to master.

File: XBMC Remote/AppDelegate.m [hunk 1]
```diff
@@ -3950,6 +3951,58 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
...
+                @"YES", @"enableCollectionView",
+                [self itemSizes_Music], @"itemSizes",
+            ],
+        ] mutableCopy];
+        
+        menu_Favourites.mainFields = @[
+            @{
+                @"itemid": @"favourites",
+                @"row1": @"title",
+                @"row2": @"none",
```
Comment (kambala-decapitator) [inline XBMC Remote/AppDelegate.m:3988]: is it possible to use `[NSNull null]` instead of `@"none"`?

File: XBMC Remote/AppDelegate.m [hunk 2]
```diff
@@ -3950,6 +3951,58 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
         @[]
     ];
     
+#pragma mark - Favourites
+        menu_Favourites.mainLabel = LOCALIZED_STR(@"Favourites");
+        menu_Favourites.upperLabel = LOCALIZED_STR(@"Choose your");
+        menu_Favourites.icon = @"icon_home_picture_alt";
+        menu_Favourites.family = 1;
```
Comment (kambala-decapitator) [inline XBMC Remote/AppDelegate.m:3958]: what is 1?

File: XBMC Remote/DetailViewController.m [hunk 3]
```diff
@@ -1204,6 +1204,24 @@ -(void)didSelectItemAtIndexPath:(NSIndexPath *)indexPath item:(NSDictionary *)it
...
+                       failure: NSLocalizedString(@"Unable to activate the window", nil)
+             ];
+        }
+        // Selected favourite item is a media type -> play it
+        else if ([item[@"type"] isEqualToString:@"media"]) {
+            [self playerOpen: @{@"item": @{@"file": item[@"path"] }} index:nil];
+        }
+        // Selected favourite item is an unknown type -> throw an error
+        else {
+            [messagesView showMessage:NSLocalizedString(@"Cannot do that", nil) timeout:2.0 color:[Utilities getSystemRed:0.95]];
```
Comment (kambala-decapitator) [inline XBMC Remote/DetailViewController.m:1222]: I'd suggest to show `item` info in the dialog, at least the type

File: XBMC Remote/Utilities.m [hunk 4]
```diff
@@ -656,7 +656,12 @@ + (NSString*)getThumbnailFromDictionary:(NSDictionary*)dict useBanner:(BOOL)useB
 + (NSString*)formatStringURL:(NSString*)path serverURL:(NSString*)serverURL {
     NSString *urlString = @"";
     if (path.length > 0 && ![path isEqualToString:@"(null)"]) {
-        urlString = [NSString stringWithFormat:@"http://%@%@", serverURL, [path stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLPathAllowedCharacterSet]]];
+        if ([path rangeOfString:@"image://"].location == NSNotFound) {
```
Comment (kambala-decapitator) [inline XBMC Remote/Utilities.m:659]: probably `hasPrefix` method makes more sense here?

File: XBMC Remote/Utilities.m [hunk 5]
```diff
@@ -656,7 +656,12 @@ + (NSString*)getThumbnailFromDictionary:(NSDictionary*)dict useBanner:(BOOL)useB
...
+            urlString = path;
+        }
+        else {
+            urlString = [NSString stringWithFormat:@"http://%@%@", serverURL, [path stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLPathAllowedCharacterSet]]];
```
Comment (kambala-decapitator) [inline XBMC Remote/Utilities.m:663]: don't you want to strip `image://` from the URL?

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: XBMC Remote/AppDelegate.m (hunk 2 above)
Comment (wutschel) [reply XBMC Remote/AppDelegate.m:3958]: Oh, this is a magic number which is obviously used when determining which type the controller is that shall be loaded. 1 for `DetailViewController`, 2 for `NowPlaying`, 3 for `RemoteController` and 4 for `HostManagementViewController`. I was not bothering to change this.

File: XBMC Remote/AppDelegate.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/AppDelegate.m:3988]: Yes, this is possible. Will change this for the next update.

File: XBMC Remote/DetailViewController.m (hunk 3 above)
Comment (wutschel) [reply XBMC Remote/DetailViewController.m:1222]: No problem, will be part of next update.

File: XBMC Remote/Utilities.m (hunk 4 above)
Comment (wutschel) [reply XBMC Remote/Utilities.m:659]: Thanks, I will change this to `if (![path hasPrefix:@"image://"]) {`

File: XBMC Remote/Utilities.m (hunk 5 above)
Comment (wutschel) [reply XBMC Remote/Utilities.m:663]: No, it seems this is needed to load the image which are in Kodi´s thumbnail path. If I remove this, the thumbs are not loaded anymore.

File: XBMC Remote/AppDelegate.m (hunk 2 above)
Comment (wutschel) [reply XBMC Remote/AppDelegate.m:3958]: I could change this to use a enum in separate PR.

**Comment by wutschel** [comment]
Rebased to master and reworked according to review comments. Changing `family` to enum can be done in separate PR.

**Comment by kambala-decapitator** [comment]
> Changing `family` to enum can be done in separate PR.

sure, that's fine

**Review by kambala-decapitator — APPROVED** [review]

**Comment by wutschel** [comment]
Resolved conflicts caused by making main menu configurable. Rebased to master and ready to go.

**Pull request merged**